### ISSUE TITLE
paddle.jit.save无限吃内存

### ISSUE THREAD

**Issue by ZSharp7** [PaddlePaddle/Paddle#49715]
### 请提出你的问题 Please ask your question

paddlepaddle: v2.4.1
我在将paddle模型使用"paddle.jit.save"的过程中，一直无限吃内存, 但是未报错，最后占满机器内存被系统killed。是哪里我使用不当吗？
基于：https://github.com/geoyee/Imatch-P

jit.save代码：
```
    superglue = SuperGlue()
    superglue_state_dict = paddle.load("models/weights/superglue_outdoor.pdparams")
    superglue.set_state_dict(superglue_state_dict)
    superglue.eval()
    net = to_static(superglue,
                    input_spec=[
                        InputSpec(shape=[1, 256, -1], dtype='float32', name='descriptors0'),
                        InputSpec(shape=[1, 256, -1], dtype='float32', name='descriptors1'),
                        InputSpec(shape=[1, -1, 2], dtype='float32', name='keypoints0'),
                        InputSpec(shape=[1, -1, 2], dtype='float32', name='keypoints1'),
                        InputSpec(shape=[4], dtype='float32', name='image0_shape'),
                        InputSpec(shape=[4], dtype='float32', name='image1_shape'),
                        InputSpec(shape=[1, -1], dtype='float32', name='scores0'),
                        InputSpec(shape=[1, -1], dtype='float32', name='scores1')
                    ])
    paddle.jit.save(net, 'superglue/superglue')
```
模型源码：
```python
from copy import deepcopy
import os.path as osp
import paddle
import paddle.nn as nn
import paddle.nn.functional as F


def MLP(channels, do_bn=True):
    """ Multi-layer perceptron """
    n = len(channels)
    layers = []
    for i in range(1, n):
        layers.append(
            nn.Conv1D(channels[i - 1], channels[i], kernel_size=1, bias_attr=True))
        if i < (n-1):
            if do_bn:
                layers.append(nn.BatchNorm1D(channels[i]))
            layers.append(nn.ReLU())
    return nn.Sequential(*layers)


def normalize_keypoints(kpts, image_shape):
    """ Normalize keypoints locations based on image image_shape"""
    _, _, height, width = image_shape
    one = paddle.to_tensor(1).astype(kpts.dtype)
    size = paddle.stack([one * width, one * height])[None]
    center = size / 2
    scaling = size.max(1, keepdim=True) * 0.7
    return (kpts - center[:, None, :].squeeze(-1)) / scaling[:, None, :].squeeze(-1)


class KeypointEncoder(nn.Layer):
    """ Joint encoding of visual appearance and location using MLPs"""
    def __init__(self, feature_dim, layers):
        super().__init__()
        self.encoder = MLP([3] + layers + [feature_dim])
        nn.initializer.Constant(value=0.0)(self.encoder[-1].bias)

    def forward(self, kpts, scores):
        inputs = [kpts.transpose((0, 2, 1)), scores.unsqueeze(1)]
        return self.encoder(paddle.concat(inputs, axis=1))


def attention(query, key, value):
    dim = query.shape[1]
    scores = paddle.einsum("bdhn,bdhm->bhnm", query, key) / dim ** .5
    prob = F.softmax(scores, axis=-1)
    return paddle.einsum("bhnm,bdhm->bdhn", prob, value), prob


class MultiHeadedAttention(nn.Layer):
    """ Multi-head attention to increase model expressivitiy """
    def __init__(self, num_heads, d_model):
        super().__init__()
        assert d_model % num_heads == 0
        self.dim = d_model // num_heads
        self.num_heads = num_heads
        self.merge = nn.Conv1D(d_model, d_model, kernel_size=1)
        self.proj = nn.LayerList([deepcopy(self.merge) for _ in range(3)])

    def forward(self, query, key, value):
        batch_dim = query.shape[0]
        query, key, value = [l(x).reshape((batch_dim, self.dim, self.num_heads, -1))
                             for l, x in zip(self.proj, (query, key, value))]
        x, _ = attention(query, key, value)
        return self.merge(x.reshape((batch_dim, self.dim*self.num_heads, -1)))


class AttentionalPropagation(nn.Layer):
    def __init__(self, feature_dim: int, num_heads: int):
        super().__init__()
        self.attn = MultiHeadedAttention(num_heads, feature_dim)
        self.mlp = MLP([feature_dim*2, feature_dim*2, feature_dim])
        nn.initializer.Constant(value=0.0)(self.mlp[-1].bias)

    def forward(self, x, source):
        message = self.attn(x, source, source)
        return self.mlp(paddle.concat([x, message], axis=1))


class AttentionalGNN(nn.Layer):
    def __init__(self, feature_dim, layer_names):
        super().__init__()
        self.layers = nn.LayerList([
            AttentionalPropagation(feature_dim, 4)
            for _ in range(len(layer_names))])
        self.names = layer_names

    def forward(self, desc0, desc1):
        for layer, name in zip(self.layers, self.names):
            if name == "cross":
                src0, src1 = desc1, desc0
            else:  # if name == "self":
                src0, src1 = desc0, desc1
            delta0, delta1 = layer(desc0, src0), layer(desc1, src1)
            desc0, desc1 = (desc0 + delta0), (desc1 + delta1)
        return desc0, desc1


def log_sinkhorn_iterations(Z, log_mu, log_nu, iters):
    """ Perform Sinkhorn Normalization in Log-space for stability"""
    u, v = paddle.zeros_like(log_mu), paddle.zeros_like(log_nu)
    for _ in range(iters):
        u = log_mu - paddle.logsumexp(Z + v.unsqueeze(1), axis=2)
        v = log_nu - paddle.logsumexp(Z + u.unsqueeze(2), axis=1)
    return Z + u.unsqueeze(2) + v.unsqueeze(1)


def log_optimal_transport(scores, alpha, iters):
    """ Perform Differentiable Optimal Transport in Log-space for stability"""
    b, m, n = scores.shape
    one = paddle.to_tensor(1).astype(scores.dtype)
    ms, ns = (m * one).astype(scores.dtype), (n * one).astype(scores.dtype)
    bins0 = alpha.expand((b, m, 1))
    bins1 = alpha.expand((b, 1, n))
    alpha = alpha.expand((b, 1, 1))
    couplings = paddle.concat([paddle.concat([scores, bins0], -1),
                               paddle.concat([bins1, alpha], -1)], 1)
    norm = -(ms + ns).log()
    log_mu = paddle.concat([norm.expand((m, )), ns.log() + norm])
    log_nu = paddle.concat([norm.expand((n, )), ms.log() + norm])
    log_mu = log_mu[None].expand((b, -1))
    log_nu = log_nu[None].expand((b, -1))
    Z = log_sinkhorn_iterations(couplings, log_mu, log_nu, iters)
    Z = Z - norm  # multiply probabilities by M+N
    return Z


def arange_like(x, dim):
    return paddle.to_tensor(list(range(x.shape[dim]))).astype(x.dtype)


def paddle_gather(x, dim, index):
    index_shape = index.shape
    index_flatten = index.flatten()
    if dim < 0:
        dim = len(x.shape) + dim
    nd_index = []
    for k in range(len(x.shape)):
        if k == dim:
            nd_index.append(index_flatten)
        else:
            reshape_shape = [1] * len(x.shape)
            reshape_shape[k] = x.shape[k]
            x_arange = paddle.arange(x.shape[k], dtype=index.dtype)
            x_arange = x_arange.reshape(reshape_shape)
            dim_index = paddle.expand(x_arange, index_shape).flatten()
            nd_index.append(dim_index)
    ind2 = paddle.transpose(paddle.stack(nd_index), [1, 0]).astype("int64")
    paddle_out = paddle.gather_nd(x, ind2).reshape(index_shape)
    return paddle_out


class SuperGlue(nn.Layer):
    """SuperGlue feature matching middle-end
    Given two sets of keypoints and locations, we determine the
    correspondences by:
      1. Keypoint Encoding (normalization + visual feature and location fusion)
      2. Graph Neural Network with multiple self and cross-attention layers
      3. Final projection layer
      4. Optimal Transport Layer (a differentiable Hungarian matching algorithm)
      5. Thresholding matrix based on mutual exclusivity and a match_threshold
    The correspondence ids use -1 to indicate non-matching points.
    Paul-Edouard Sarlin, Daniel DeTone, Tomasz Malisiewicz, and Andrew
    Rabinovich. SuperGlue: Learning Feature Matching with Graph Neural
    Networks. In CVPR, 2020. https://arxiv.org/abs/1911.11763
    """
    def __init__(self):
        super().__init__()
        self.kenc = KeypointEncoder(256, [32, 64, 128, 256])
        self.gnn = AttentionalGNN(feature_dim=256, layer_names=["self", "cross"] * 9)
        self.final_proj = nn.Conv1D(256, 256, kernel_size=1, bias_attr=True)
        self.sinkhorn_iterations = 20
        self.match_threshold = 0.2
        bin_score = paddle.create_parameter([1], dtype="float32",
                    default_initializer=nn.initializer.Constant(value=1.))
        self.add_parameter("bin_score", bin_score)

    def forward(self, descriptors0, descriptors1, keypoints0, keypoints1, image0_shape, image1_shape, scores0, scores1):
        """Run SuperGlue on a pair of keypoints and descriptors"""
        desc0, desc1 = descriptors0, descriptors1
        kpts0, kpts1 = keypoints0, keypoints1
        # Keypoint normalization.
        kpts0 = normalize_keypoints(kpts0, image0_shape)
        kpts1 = normalize_keypoints(kpts1, image1_shape)
        # Keypoint MLP encoder.
        desc0 = desc0 + self.kenc(kpts0, scores0)
        desc1 = desc1 + self.kenc(kpts1, scores1)
        # kpts0 = normalize_keypoints(kpts0, data["image0"].shape)
        # kpts1 = normalize_keypoints(kpts1, data["image1"].shape)
        # # Keypoint MLP encoder.
        # desc0 = desc0 + self.kenc(kpts0, data["scores0"])
        # desc1 = desc1 + self.kenc(kpts1, data["scores1"])
        # Multi-layer Transformer network.
        desc0, desc1 = self.gnn(desc0, desc1)
        # Final MLP projection.
        mdesc0, mdesc1 = self.final_proj(desc0), self.final_proj(desc1)
        # Compute matching descriptor distance.
        scores = paddle.einsum("bdn,bdm->bnm", mdesc0, mdesc1)
        scores = scores / 256 ** .5
        # Run the optimal transport.
        scores = log_optimal_transport(
            scores, self.bin_score,
            iters=self.sinkhorn_iterations)
        # Get the matches with score above "match_threshold".
        max0, max1 = scores[:, :-1, :-1].max(2), scores[:, :-1, :-1].max(1)
        indices0 = scores[:, :-1, :-1].argmax(2, keepdim=True).squeeze(2)
        indices1 = scores[:, :-1, :-1].argmax(1, keepdim=True).squeeze(1)
        mutual0 = arange_like(indices0, 1)[None] == paddle_gather(indices1, 1, indices0)
        mutual1 = arange_like(indices1, 1)[None] == paddle_gather(indices0, 1, indices1)
        zero = paddle.to_tensor(0).astype(scores.dtype)
        mscores0 = paddle.where(mutual0, max0.exp(), zero)
        mscores1 = paddle.where(mutual1, paddle_gather(mscores0, 1, indices1), zero)
        valid0 = mutual0 & (mscores0 > self.match_threshold)
        valid1 = mutual1 & paddle_gather(valid0, 1, indices1)
        indices0 = paddle.where(valid0, indices0, paddle.to_tensor(-1).astype(indices0.dtype))
        indices1 = paddle.where(valid1, indices1, paddle.to_tensor(-1).astype(indices1.dtype))
        return [
            indices0,  # use -1 for invalid match
            indices1,  # use -1 for invalid match
            mscores0,
            mscores1
        ]
```

**Comment by zhangbo9674** [comment]
你好，有相同issue可以查看一下：https://github.com/PaddlePaddle/Paddle/issues/36115

**Comment by ZSharp7** [comment]
@zhangbo9674 我这不是训练过程中每次save，这是直接将模型转成save的过程。程序会卡在 paddle.jit.save中一直运行，直到内存吃满被系统killed。 实测forward加上“@paddle.jit.to_static”无法解决问题。

**Comment by 2742195759** [comment]
这里已经帮您定位到了，属于是动转静的一个bug，这个bug我们会后续思考如何解决。您的问题主要是因为：
```
def arange_like(x, dim):
    return paddle.to_tensor(list(range(x.shape[dim]))).astype(x.dtype)
```
函数被动转静转为了
`list(_jst.Call(range)(_jst.Shape(x)[dim]))`
最后对一个Variable调用了 list(var)，导致卡死。

目前临时的解决方案可以这么修改：
```
def arange_like(x, dim):
    return paddle.arange(x.shape[dim]).astype(x.dtype)
```
就可以成功导出。

**Comment by ZSharp7** [comment]
@2742195759 感谢支持，已经能正常转出模型了。但是同样的输入，推理paddleinference的推理结果和paddle原生推理结果有很大差异，这是因为什么？
## 测试数据：
链接: https://pan.baidu.com/s/19-7-kf_n8iiIutEArtn-dQ 提取码: atj6 
## 测试代码：
```python
# -*- coding: utf-8 -*-
import numpy as np
import cv2
import paddle
from paddle.inference import Config
from paddle.inference import create_predictor
from paddle.inference import PrecisionType
from models import SuperGlue


descriptors0 = np.load("test_data/descriptors0.npy")
descriptors1 = np.load("test_data/descriptors1.npy")

scores0 = np.load("test_data/scores0.npy")
scores1 = np.load("test_data/scores1.npy")

keypoints0 = np.load("test_data/keypoints0.npy")
keypoints1 = np.load("test_data/keypoints1.npy")

image0_shape = np.load("test_data/image0_shape.npy")
image1_shape = np.load("test_data/image1_shape.npy")


def test_paddle():
    global descriptors0,descriptors1,scores0,scores1,keypoints0,keypoints1,image0_shape,image1_shape
    superglue = SuperGlue()
    superglue.eval()
    superglue_state_dict = paddle.load("models/weights/superglue_outdoor.pdparams")
    superglue.set_state_dict(superglue_state_dict)
    descriptors0_tensor = paddle.to_tensor(descriptors0, "float32")
    descriptors1_tensor = paddle.to_tensor(descriptors1, "float32")
    scores0_tensor = paddle.to_tensor(scores0, "float32")
    scores1_tensor = paddle.to_tensor(scores1, "float32")
    keypoints0_tensor = paddle.to_tensor(keypoints0, "float32")
    keypoints1_tensor = paddle.to_tensor(keypoints1, "float32")
    image0_shape_tensor = paddle.to_tensor(image0_shape, "float32")
    image1_shape_tensor = paddle.to_tensor(image1_shape, "float32")
    out = superglue(descriptors0_tensor, descriptors1_tensor, keypoints0_tensor, keypoints1_tensor, image0_shape_tensor, image1_shape_tensor, scores0_tensor, scores1_tensor)
    print(out)


def test_paddle_inference():
    global descriptors0,descriptors1,scores0,scores1,keypoints0,keypoints1,image0_shape,image1_shape

    data = {
        "descriptors0": descriptors0,
        "descriptors1": descriptors1,
        "scores0": scores0,
        "scores1": scores1,
        "keypoints0": keypoints0,
        "keypoints1": keypoints1,
        "image0_shape": image0_shape,
        "image1_shape": image1_shape
    }

    model_file = r"superglue/superglue.pdmodel"
    params_file = r"superglue/superglue.pdiparams"

    config = Config(model_file, params_file)

    # config.enable_use_gpu(1000, 2)
    config.enable_memory_optim()
    config.glog_info_disabled()
    predictor = create_predictor(config)

    input_names = predictor.get_input_names()
    for name in input_names:
        input_tensor = predictor.get_input_handle(name)
        input_tensor.reshape(data[name].shape)
        input_tensor.copy_from_cpu(data[name].copy())

    predictor.run()

    results = []
    # get out data from output tensor
    output_names = predictor.get_output_names()
    for i, name in enumerate(output_names):
        output_tensor = predictor.get_output_handle(name)
        output_data = output_tensor.copy_to_cpu()
        results.append(output_data)
    print(results)
if __name__ == '__main__':
    test_paddle()
    test_paddle_inference()
```
## 推理输出：
* paddle原生推理输出
```
[Tensor(shape=[1, 4], dtype=int64, place=Place(cpu), stop_gradient=False,
       [[0, 1, 2, 3]]), Tensor(shape=[1, 4], dtype=int64, place=Place(cpu), stop_gradient=False,
       [[0, 1, 2, 3]]), Tensor(shape=[1, 4], dtype=float32, place=Place(cpu), stop_gradient=False,
       [[0.97485626, 0.97495711, 0.97377539, 0.97495550]]), Tensor(shape=[1, 4], dtype=float32, place=Place(cpu), stop_gradient=False,
       [[0.97485626, 0.97495711, 0.97377539, 0.97495550]])]
```
* paddle_inference推理输出
```
[array([[-1,  1,  3,  2]]), array([[-1, -1, -1, -1]]), array([[0., 0., 0., 0.]], dtype=float32), array([[0., 0., 0., 0.]], dtype=float32)]
```

**Comment by 2742195759** [comment]
我本地试了一下，发现精度是可以对齐的。怀疑是你的模型是不是load路径有问题。我这边使用随机的参数（因为你没有上传参数文件），得到的结果如下：

Predictor运行的结果：
![截屏2023-01-11 20 17 57](https://user-images.githubusercontent.com/16025309/211804381-c33b0896-405d-4851-9632-2ef39e3dcda1.png)

动态图运行的结果：
![截屏2023-01-11 20 18 08](https://user-images.githubusercontent.com/16025309/211804432-8b8efca7-6158-4808-bc78-6b85d7c1ce0d.png)

我这里使用的是develop分支，也不确定是不是版本的原因，您可以排查一下。

**Comment by ZSharp7** [comment]
@2742195759 非常感谢！我试了一下develop分支，使用测试数据paddle_inference运行结果和paddle原生是一致的。这是不是要等下一个release版本出来再使用。
* paddle_inference
```
[Tensor(shape=[1, 4], dtype=int64, place=Place(cpu), stop_gradient=True,
       [[0, 1, 2, 3]]), Tensor(shape=[1, 4], dtype=int64, place=Place(cpu), stop_gradient=True,
       [[0, 1, 2, 3]]), Tensor(shape=[1, 4], dtype=float32, place=Place(cpu), stop_gradient=False,
       [[0.97477978, 0.97442311, 0.97431785, 0.97450978]]), Tensor(shape=[1, 4], dtype=float32, place=Place(cpu), stop_gradient=False,
       [[0.97477978, 0.97442311, 0.97431785, 0.97450978]])]
```
* paddle原生
```
[array([[0, 1, 2, 3]]), array([[0, 1, 2, 3]]), array([[0.9747798 , 0.9744231 , 0.97431785, 0.9745098 ]], dtype=float32), array([[0.9747798 , 0.9744231 , 0.97431785, 0.9745098 ]], dtype=float32)]
```

**Comment by 2742195759** [comment]
应该是了，可以试试最新的2.4版本是否可以。如果也不可以就等下一个版本吧，应该是Inference某些bug在develop中被修复了。

**Comment by ZSharp7** [comment]
@2742195759 好的，当前使用的就是paddlepaddle: v2.4.1版本。
